### PR TITLE
Added quirk for TS0601 _TZE204_p3lqqy2r

### LINKED ISSUE #2868: [Device Support Request] Namron Zigbee Touch Thermostat 16A 4512752/4512753
### Problem description

New Namron Zigbee Touch Termostat 2.0 is identified as TS0601 by _TZE204_p3lqqy2r with no controls or diagnostic.

https://www.namron.com/products/namron-zigbee-touch-termostat-16a-sort-41/ / 4512753
https://www.namron.com/products/namron-zigbee-touch-termostat-16a-hvit-23/ / 4512752

### Solution description

The device should have similar support as it's predecessor https://www.namron.com/products/namron-zigbee-touch-termostat-16a-sort / 4512738.

### Screenshots/Video

<details><summary>Screenshots/Video</summary>

image

</details>

### Device signature

<details><summary>Device signature</summary>

```json
{
  "node_descriptor": "NodeDescriptor(logical_type=<LogicalType.Router: 1>, complex_descriptor_available=0, user_descriptor_available=0, reserved=0, aps_flags=0, frequency_band=<FrequencyBand.Freq2400MHz: 8>, mac_capability_flags=<MACCapabilityFlags.FullFunctionDevice|MainsPowered|RxOnWhenIdle|AllocateAddress: 142>, manufacturer_code=4098, maximum_buffer_size=82, maximum_incoming_transfer_size=82, server_mask=11264, maximum_outgoing_transfer_size=82, descriptor_capability_field=<DescriptorCapability.NONE: 0>, *allocate_address=True, *is_alternate_pan_coordinator=False, *is_coordinator=False, *is_end_device=False, *is_full_function_device=True, *is_mains_powered=True, *is_receiver_on_when_idle=True, *is_router=True, *is_security_capable=False)",
  "endpoints": {
    "1": {
      "profile_id": "0x0104",
      "device_type": "0x0051",
      "input_clusters": [
        "0x0000",
        "0x0004",
        "0x0005",
        "0xef00"
      ],
      "output_clusters": [
        "0x000a",
        "0x0019"
      ]
    }
  },
  "manufacturer": "_TZE204_p3lqqy2r",
  "model": "TS0601",
  "class": "zigpy.device.Device"
}
```

</details>

### Diagnostic information

<details><summary>Diagnostic information</summary>

```json
{
  "home_assistant": {
    "installation_type": "Home Assistant OS",
    "version": "2023.12.4",
    "dev": false,
    "hassio": true,
    "virtualenv": false,
    "python_version": "3.11.6",
    "docker": true,
    "arch": "x86_64",
    "timezone": "Europe/Oslo",
    "os_name": "Linux",
    "os_version": "6.1.63-haos",
    "supervisor": "2023.12.0",
    "host_os": "Home Assistant OS 11.2",
    "docker_version": "24.0.7",
    "chassis": "vm",
    "run_as_root": true
  },
  "custom_components": {
    "hacs": {
      "version": "1.33.0",
      "requirements": [
        "aiogithubapi>=22.10.1"
      ]
    },
    "easee": {
      "version": "0.9.55",
      "requirements": [
        "pyeasee==0.7.55"
      ]
    },
    "tesla_custom": {
      "version": "3.19.5",
      "requirements": [
        "teslajsonpy==3.9.8"
      ]
    }
  },
  "integration_manifest": {
    "domain": "zha",
    "name": "Zigbee Home Automation",
    "after_dependencies": [
      "onboarding",
      "usb"
    ],
    "codeowners": [
      "@dmulcahey",
      "@adminiuga",
      "@puddly",
      "@TheJulianJES"
    ],
    "config_flow": true,
    "dependencies": [
      "file_upload"
    ],
    "documentation": "https://www.home-assistant.io/integrations/zha",
    "iot_class": "local_polling",
    "loggers": [
      "aiosqlite",
      "bellows",
      "crccheck",
      "pure_pcapy3",
      "zhaquirks",
      "zigpy",
      "zigpy_deconz",
      "zigpy_xbee",
      "zigpy_zigate",
      "zigpy_znp",
      "universal_silabs_flasher"
    ],
    "requirements": [
      "bellows==0.37.4",
      "pyserial==3.5",
      "pyserial-asyncio==0.6",
      "zha-quirks==0.0.108",
      "zigpy-deconz==0.22.3",
      "zigpy==0.60.2",
      "zigpy-xbee==0.20.1",
      "zigpy-zigate==0.12.0",
      "zigpy-znp==0.12.1",
      "universal-silabs-flasher==0.0.15",
      "pyserial-asyncio-fast==0.11"
    ],
    "usb": [
      {
        "vid": "10C4",
        "pid": "EA60",
        "description": "*2652*",
        "known_devices": [
          "slae.sh cc2652rb stick"
        ]
      },
      {
        "vid": "1A86",
        "pid": "55D4",
        "description": "*sonoff*plus*",
        "known_devices": [
          "sonoff zigbee dongle plus v2"
        ]
      },
      {
        "vid": "10C4",
        "pid": "EA60",
        "description": "*sonoff*plus*",
        "known_devices": [
          "sonoff zigbee dongle plus"
        ]
      },
      {
        "vid": "10C4",
        "pid": "EA60",
        "description": "*tubeszb*",
        "known_devices": [
          "TubesZB Coordinator"
        ]
      },
      {
        "vid": "1A86",
        "pid": "7523",
        "description": "*tubeszb*",
        "known_devices": [
          "TubesZB Coordinator"
        ]
      },
      {
        "vid": "1A86",
        "pid": "7523",
        "description": "*zigstar*",
        "known_devices": [
          "ZigStar Coordinators"
        ]
      },
      {
        "vid": "1CF1",
        "pid": "0030",
        "description": "*conbee*",
        "known_devices": [
          "Conbee II"
        ]
      },
      {
        "vid": "0403",
        "pid": "6015",
        "description": "*conbee*",
        "known_devices": [
          "Conbee III"
        ]
      },
      {
        "vid": "10C4",
        "pid": "8A2A",
        "description": "*zigbee*",
        "known_devices": [
          "Nortek HUSBZB-1"
        ]
      },
      {
        "vid": "0403",
        "pid": "6015",
        "description": "*zigate*",
        "known_devices": [
          "ZiGate+"
        ]
      },
      {
        "vid": "10C4",
        "pid": "EA60",
        "description": "*zigate*",
        "known_devices": [
          "ZiGate"
        ]
      },
      {
        "vid": "10C4",
        "pid": "8B34",
        "description": "*bv 2010/10*",
        "known_devices": [
          "Bitron Video AV2010/10"
        ]
      }
    ],
    "zeroconf": [
      {
        "type": "_esphomelib._tcp.local.",
        "name": "tube*"
      },
      {
        "type": "_zigate-zigbee-gateway._tcp.local.",
        "name": "*zigate*"
      },
      {
        "type": "_zigstar_gw._tcp.local.",
        "name": "*zigstar*"
      },
      {
        "type": "_uzg-01._tcp.local.",
        "name": "uzg-01*"
      },
      {
        "type": "_slzb-06._tcp.local.",
        "name": "slzb-06*"
      }
    ],
    "is_built_in": true
  },
  "data": {
    "ieee": "**REDACTED**",
    "nwk": 31358,
    "manufacturer": "_TZE204_p3lqqy2r",
    "model": "TS0601",
    "name": "_TZE204_p3lqqy2r TS0601",
    "quirk_applied": false,
    "quirk_class": "zigpy.device.Device",
    "quirk_id": null,
    "manufacturer_code": 4098,
    "power_source": "Mains",
    "lqi": 144,
    "rssi": -64,
    "last_seen": "2023-12-28T15:46:27",
    "available": true,
    "device_type": "Router",
    "signature": {
      "node_descriptor": "NodeDescriptor(logical_type=<LogicalType.Router: 1>, complex_descriptor_available=0, user_descriptor_available=0, reserved=0, aps_flags=0, frequency_band=<FrequencyBand.Freq2400MHz: 8>, mac_capability_flags=<MACCapabilityFlags.FullFunctionDevice|MainsPowered|RxOnWhenIdle|AllocateAddress: 142>, manufacturer_code=4098, maximum_buffer_size=82, maximum_incoming_transfer_size=82, server_mask=11264, maximum_outgoing_transfer_size=82, descriptor_capability_field=<DescriptorCapability.NONE: 0>, *allocate_address=True, *is_alternate_pan_coordinator=False, *is_coordinator=False, *is_end_device=False, *is_full_function_device=True, *is_mains_powered=True, *is_receiver_on_when_idle=True, *is_router=True, *is_security_capable=False)",
      "endpoints": {
        "1": {
          "profile_id": "0x0104",
          "device_type": "0x0051",
          "input_clusters": [
            "0x0000",
            "0x0004",
            "0x0005",
            "0xef00"
          ],
          "output_clusters": [
            "0x000a",
            "0x0019"
          ]
        }
      },
      "manufacturer": "_TZE204_p3lqqy2r",
      "model": "TS0601"
    },
    "active_coordinator": false,
    "entities": [],
    "neighbors": [
      {
        "device_type": "Coordinator",
        "rx_on_when_idle": "On",
        "relationship": "Parent",
        "extended_pan_id": "**REDACTED**",
        "ieee": "**REDACTED**",
        "nwk": "0x0000",
        "permit_joining": "Unknown",
        "depth": "0",
        "lqi": "124"
      },
      {
        "device_type": "Router",
        "rx_on_when_idle": "On",
        "relationship": "Sibling",
        "extended_pan_id": "**REDACTED**",
        "ieee": "**REDACTED**",
        "nwk": "0x6E66",
        "permit_joining": "Unknown",
        "depth": "15",
        "lqi": "107"
      },
      {
        "device_type": "Router",
        "rx_on_when_idle": "On",
        "relationship": "Sibling",
        "extended_pan_id": "**REDACTED**",
        "ieee": "**REDACTED**",
        "nwk": "0x9F35",
        "permit_joining": "Unknown",
        "depth": "15",
        "lqi": "67"
      }
    ],
    "routes": [
      {
        "dest_nwk": "0x0000",
        "route_status": "Active",
        "memory_constrained": true,
        "many_to_one": true,
        "route_record_required": false,
        "next_hop": "0x6E66"
      }
    ],
    "endpoint_names": [
      {
        "name": "SMART_PLUG"
      }
    ],
    "user_given_name": "livingroom_thermostat_namron",
    "device_reg_id": "0d43525102b95a56bf9b7159ec66d547",
    "area_id": "living_room_1st_floor",
    "cluster_details": {
      "1": {
        "device_type": {
          "name": "SMART_PLUG",
          "id": 81
        },
        "profile_id": 260,
        "in_clusters": {
          "0x0000": {
            "endpoint_attribute": "basic",
            "attributes": {
              "0x0001": {
                "attribute_name": "app_version",
                "value": 70
              },
              "0x0003": {
                "attribute_name": "hw_version",
                "value": 1
              },
              "0x0004": {
                "attribute_name": "manufacturer",
                "value": "_TZE204_p3lqqy2r"
              },
              "0x0005": {
                "attribute_name": "model",
                "value": "TS0601"
              }
            },
            "unsupported_attributes": {
              "0x4000": {
                "attribute_name": "sw_build_id"
              }
            }
          },
          "0x0004": {
            "endpoint_attribute": "groups",
            "attributes": {},
            "unsupported_attributes": {}
          },
          "0x0005": {
            "endpoint_attribute": "scenes",
            "attributes": {},
            "unsupported_attributes": {}
          },
          "0xef00": {
            "endpoint_attribute": null,
            "attributes": {},
            "unsupported_attributes": {}
          }
        },
        "out_clusters": {
          "0x0019": {
            "endpoint_attribute": "ota",
            "attributes": {},
            "unsupported_attributes": {}
          },
          "0x000a": {
            "endpoint_attribute": "time",
            "attributes": {},
            "unsupported_attributes": {}
          }
        }
      }
    }
  }
}
```

</details>

### Logs

<details><summary>Logs</summary>

```python
[Paste the logs here]
```

</details>

### Custom quirk

<details><summary>Custom quirk</summary>

```python
[Paste your custom quirk here]
```

</details>

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by mellowism**
## Proposed change
Added quirk for TS0601 _TZE204_p3lqqy2r

## Additional information
Fixes #2868 

## Checklist

- [x] The changes are tested and work correctly
- [ ] `pre-commit` checks pass / the code has been formatted using Black
- [ ] Tests have been added to verify that the new code works

### CODE REVIEW

File: zhaquirks/tuya/TS0601_thermostat [hunk 1]
```diff
@@ -0,0 +1,144 @@
+from zigpy.quirks import CustomDevice
+from zigpy.types import enum8, int16s
```
Comment (prairiesnpr) [inline zhaquirks/tuya/TS0601_thermostat:2]: From a style standpoint, the other quirks use `import zigpy.types as t`

File: zhaquirks/tuya/TS0601_thermostat [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+    Floor = 0x01
+    Both = 0x02
+
+
+class RunningState(enum8):
+    Idle = 0x00
+    Heat = 0x01
+
+
+quirk = (
```
Comment (prairiesnpr) [inline zhaquirks/tuya/TS0601_thermostat:31]: `quirk` isn't used, you can remove `quirk = `

File: zhaquirks/tuya/TS0601_thermostat [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+    TuyaQuirkBuilder()
+    .applies_to("_TZE204_p3lqqy2r", "TS0601")
```
Comment (prairiesnpr) [inline zhaquirks/tuya/TS0601_thermostat:33]: This should just be `TuyaQuirkBuilder("_TZE204_p3lqqy2r", "TS0601")`

File: zhaquirks/tuya/TS0601_thermostat [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+        attribute_name="energy",
+        type=int16s,
+        device_class=SensorDeviceClass.ENERGY,
+        state_class=SensorStateClass.TOTAL,
+        unit=UnitOfEnergy.KILO_WATT_HOUR,
+        translation_key="energy",
+        fallback_name="Energy",
+    )
+    .tuya_sensor(
+        dp_id=122,
```
Comment (prairiesnpr) [inline zhaquirks/tuya/TS0601_thermostat:128]: Please be consistent with `dp_id`, let's not mix hexadecimal and decimal representations.

File: zhaquirks/tuya/TS0601_thermostat [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+class RunningState(enum8):
+    Idle = 0x00
+    Heat = 0x01
+
+
+quirk = (
+    TuyaQuirkBuilder()
+    .applies_to("_TZE204_p3lqqy2r", "TS0601")
+    
+    # Add climate attributes
```
Comment (prairiesnpr) [inline zhaquirks/tuya/TS0601_thermostat:35]: This comment isn't required.

File: zhaquirks/tuya/TS0601_thermostat [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+    .tuya_number(
+        dp_id=0x10,
+        type=int16s,
+        attribute_name="current_heating_setpoint",
+        min_value=5,
+        max_value=35,
+        step=1,
+        unit=UnitOfTemperature.CELSIUS,
+        translation_key="heating_setpoint",
+        fallback_name="Heating Setpoint",
```
Comment (prairiesnpr) [inline zhaquirks/tuya/TS0601_thermostat:59]: Fallback names should have the first word capitalized and then be lowercase.

**Comment by raphaelc75** [comment]
I have tested this quirk, and it looks really nice. 
Just one comment: 
It seems like the value of Energy [kWh] should be divided by 100 or something like that.

**Comment by prairiesnpr** [comment]
> I have tested this quirk, and it looks really nice. Just one comment: It seems like the value of Energy [kWh] should be divided by 100 or something like that.

You can scale the value by adding `converter=lambda x: x * 0.01,` to the appropriate `.tuya_sensor`

**Comment by raphaelc75** [comment]
Thank you! 
I added your code at line 126, and now the Energy value seems to be correct :-) 

>     .tuya_sensor(
>        dp_id=123,
>        attribute_name="energy",
>        type=int16s,
>        device_class=SensorDeviceClass.ENERGY,
>        state_class=SensorStateClass.TOTAL,
>        unit=UnitOfEnergy.KILO_WATT_HOUR,
>        translation_key="energy",
>        fallback_name="Energy",
>        converter=lambda x: x * 0.01,
>        )

**Comment by mellowism** [comment]
I'm not really a developer, and this is my first-ever pull request, and it was done with the help of an LLM. However, it works, and I wanted to share it with others who have purchased the same device. If anyone with more experience could assist with the proposed changes, that would be fantastic. Additionally, if there's a way to add a climate entity, that would be amazing. I spent over a day trying to achieve that...

**Comment by prairiesnpr** [comment]
> if there's a way to add a climate entity, that would be amazing. I spent over a day trying to achieve that...

I don't have any Zigbee thermostats, let alone this model, but you can try this quirk.
https://gist.github.com/prairiesnpr/1b17a42135637a5aa43ce7330837c2a9

If it doesn't work, set debug logging to true and send me the logs.

**Comment by mellowism** [comment]
> > if there's a way to add a climate entity, that would be amazing. I spent over a day trying to achieve that...
> 
> I don't have any Zigbee thermostats, let alone this model, but you can try this quirk. https://gist.github.com/prairiesnpr/1b17a42135637a5aa43ce7330837c2a9
> 
> If it doesn't work, set debug logging to true and send me the logs.

I spent some more time trying to get that to work, but I couldn't. I've attached the logs, and I really appreciate you helping me out. 
[home-assistant_zha_2024-12-18T10-51-00.745Z.log](https://github.com/user-attachments/files/18179968/home-assistant_zha_2024-12-18T10-51-00.745Z.log)

**Comment by prairiesnpr** [comment]
> > > if there's a way to add a climate entity, that would be amazing. I spent over a day trying to achieve that...
> > 
> > 
> > I don't have any Zigbee thermostats, let alone this model, but you can try this quirk. https://gist.github.com/prairiesnpr/1b17a42135637a5aa43ce7330837c2a9
> > If it doesn't work, set debug logging to true and send me the logs.
> 
> I spent some more time trying to get that to work, but I couldn't. I've attached the logs, and I really appreciate you helping me out. [home-assistant_zha_2024-12-18T10-51-00.745Z.log](https://github.com/user-attachments/files/18179968/home-assistant_zha_2024-12-18T10-51-00.745Z.log)

Can you provide more insight on how it didn't work and also the ieee/nwk of the device? If you have a list of tests you performed that failed, knowing the timestamp is also very helpful.

**Comment by mellowism** [comment]
> Can you provide more insight on how it didn't work and also the ieee/nwk of the device? If you have a list of tests you performed that failed, knowing the timestamp is also very helpful.

In this test instance, I only have this Zigbee device, so the logs only reflect that in terms of Zigbee devices. The quirk I am using here is the one I submitted. It works fine, and all entities are working and updating. However, there is no climate entity, and there appears to be a formatting error in the Energy value.

IEEE: 8c:65:a3:ff:fe:55:36:fd
Nwk: 0xf1d6

I restarted the system at approximately 11:48 and added the device immediately afterward by enabling pairing in ZHA and putting the device in pairing mode (I assume the 11:48:47.906 timestamp corresponds to this).

Next, I pressed the arrow up on the device a couple of times to increase the temperature, then pressed down. I also cycled through the different modes by clicking M on the device: Home, Away, and Touch.

I hope this information helps. If needed, I’m happy to do a session over Discord or a similar platform. I’ve invested so much into this that I’m eager to get it working! 😊

**Comment by prairiesnpr** [comment]
> The quirk I am using here is the one I submitted. It works fine, and all entities are working and updating. However, there is no climate entity, and there appears to be a formatting error in the Energy value.

Ah, that makes sense. To get a climate entity, we need to expose that, not the individual sensors. I was looking for logs from my attempt to expose a thermostat cluster in https://gist.github.com/prairiesnpr/1b17a42135637a5aa43ce7330837c2a9 

Please attempt to use that quirk and record the logs. Wanted to confirm the quirk was loading in the first place, and then look at the results if so. Feel free to reach out on discord if that's easier.

**Pull request closed without merging**